### DISCUSSION TITLE
BC Media Format - Streaming audio to the cameras

### DISCUSSION THREAD

**Discussion by entropin** [thirtythreeforty/neolink#187] [Q&A] [answered]
Hi, Am working towards adding  "TalkAbility" functionality to Neolink, but am having a problem getting my camera to accept the audio bites am sending @QuantumEntangledAndy has helped me out quite a bit, but I wanted to see if anyone else got some insights regarding the audio recordings.

This is the `<TalkConfig>` that my code and the client is sending to the camera:
```
<?xml version="1.0" encoding="UTF-8" ?>
<body>
<TalkConfig version="1.1">
<channelId>0</channelId>
<duplex>FDX</duplex>
<audioStreamMode>followVideoStream</audioStreamMode>
<audioConfig>
<audioType>adpcm</audioType>
<sampleRate>16000</sampleRate>
<samplePrecision>16</samplePrecision>
<lengthPerEncoder>1024</lengthPerEncoder>
<soundTrack>mono</soundTrack>
</audioConfig>
</TalkConfig>
</body>
```

Here is a short WireShark capture of me streaming to the camera via the old Reolink client.
https://drive.google.com/file/d/14256PW40V2ZrgIHPISStMmbYJV0gOkoY/view?usp=sharing

The packages that the client is sending are in BC Media format, like how the camera is sending video to the client. So the most likely is that i need to encode by audio to DVI-4 adpcm.

This is how I convert the audio am working with right now `ffmpeg -i bensound-dubstep.mp3 -f wav -ar 16000 -ac 1 -acodec adpcm_ms OUTPUT.wav `

[Wikipedia ](https://en.wikipedia.org/wiki/FFmpeg)list that ffmpeg has DVI support but i cant find any information that i can understand on how to specify DVI4. Dose someone know more about this?

Finally am also am confused about this part of the protocol documentation: 

`### ADPCM

After the ADPCM header is another header of the form

- 2 bytes Magic either 0x00 0x01 or 0x00 0x7a
- 2 Bytes DVI-4 Block size
- `
What is a Block in this context? I cant find any good information about Block size and audio encoding :/

Any and all insight is welcomed, right now it feels like am just throwing random bits at the camera haha

**Comment by QuantumEntangledAndy** [comment, marked as answer]
Ok so I do not believe ffmpeg supports dvi4 adpcm (it is not the same as DVI movies). You can check which ones it supports with assuming linux/osx term with `grep`

```bash
ffmpeg -codecs | grep adpcm
```

The closest format that ffmpeg supports is the IMA one which is the same as DVI but with a different block header.

So few things support it that we created a adpcm decoder in rust `src/bc_protocol/adpcm.rs`. It may be necessary to write the encoder then you can just use more standard formats. 

It is definitely NOT `adpcm_ms`

---

Now about the block part

A block is a small chunk of audio. Every byte in adpcm depends on the previous byte. So what do you do if you want to start playing at 80% through the track? Do you load every byte before? No instead you only need to go back to the beginning of the block. In the block header is effectively a checkpoint saying this byte is this value no need to look further backwards to calculate it.

How often this checkpoint is made depends on how the adpcm was implemented and may be configurable. You may want to read our discussion in #68 this is where we actually reverse engineered this stuff and I learnt all I know about ADPCM during that PR is the BC Media format.

---


@twistedddx I remember you are good at this sort of less common audio formats do you have any recommended way to encode in DVI4 adpcm?

**Reply by entropin** [reply]
Ok, if it's that unusable, maybe the android client will use a different codec I'll try to sniff that if i can find a old APK not using AES

**Reply by QuantumEntangledAndy** [reply]
For now you might just want to focus on getting the other parts like the BC Media format sorted. 

Ignore the exact codec and just feed it any old bytes maybe the ima one. It will probably just sound like static on the camera but at least you know it will be playing.

**Reply by QuantumEntangledAndy** [reply]
I just remembered that gstreamer has some support.  Check that out [here](https://gstreamer.freedesktop.org/documentation/adpcmenc/index.html)

**Reply by QuantumEntangledAndy** [reply]
Here try this one for the gstreamer convert

```bash
gst-launch-1.0 filesrc location=nero.aac ! decodebin ! audio/x-raw ! audioconvert ! "audio/x-raw,channels=1" ! queue  ! adpcmenc blockalign=1024 layout=dvi ! filesink location=out.adpcm
```

**Reply by entropin** [reply]
blockalign in this case is the size of the blocks?

**Reply by QuantumEntangledAndy** [reply]
Yes it is the block size

**Reply by QuantumEntangledAndy** [reply]
The block size in the wireshark you sent me is also 1024 so maybe just stick with that

**Reply by entropin** [reply]
Oki, sorry I feel so uninformed right now! I think i need to go on an a side quest to learn some basics about audio codecs.. 

But are this man talking about the same codec that we are?

https://github.com/pdeljanov/Symphonia/issues/41

If so, this might help us since it's rust, but also he mentioned that ffmpeg do seam to support it?

**Reply by QuantumEntangledAndy** [reply]
It's possible that its not called dvi in ffmpeg and its one of the ima ones like `adpcm_ima_qt` or something, I don't think I tried all of them. Ima and DVI are very similar. Only in the header is their a difference where one of them stores a u16 byte of the audio data and the other stores 2 u8 bytes of the decoder state that can be used to calculate the u16 byte of audio data.

**Reply by twistedddx** [reply]
My read of the packet capture is likely exactly what everyone else already saw but just to put it in front of us:

Capture uses 01wb. We assume the data is DVI4 adpcm, reinforced by XML description also mentioning adpcm.
Swann used 01wb which was observed to be BC media magic for DVI4 adpcm
QCam used 03wb which was observed to be BC media magic for DVI4 adpcm

Capture shows size is 520 bytes. First 4 bytes is second magic, 4 bytes is previous decode state and 512 bytes of audio.
Swann size was 248 bytes. First 4 bytes is second magic, 4 bytes is previous decode state and 240 bytes of audio.
QCam size was 168 bytes. First 4 bytes is second magic, 4 bytes is previous decode state and 160 bytes of audio.

Capture shows second magic of 00 01 02 00. While 02 00 can give 512 it breaks from the traditions of little endian, half the DVI4 size and DVI4 size not audio only size.
Swann cameras had second magic of 00 01 7A 00 with 7A being 122 which is half of the DVI4 244 bytes.
QCam cameras had second magic of 00 01 52 00 with 52 being 82 which is half of the DVI4 164 bytes.

Capture shows first 01wb for each packet having decode state reset to 00 00 00 00.
01wb within a packet does appear to show storage of the decoder state.
Maybe the talk button was only pressed for a moment each time?


512 bytes of audio will give 1024 samples.
512 is also a shift from the multiples of 80 that HiSilicon suggested were common lengths for DVI4 ADPCM and was in use for other BC media examples.

**Reply by entropin** [reply]
FYI: The dump in the OP is a E1 Pro

**Reply by twistedddx** [reply]
Isn't the DVI4 block size for this data 516?

**Reply by twistedddx** [reply]
In #68 we found that for Swann this code for gstreamer was able to decode the data from a Swann camera.

Some(StreamFormat::ADPCM) => "caps=audio/x-adpcm,layout=dvi,block_align=244,channels=1,rate=8000 ! queue silent=true max-size-bytes=10485760 min-threshold-bytes=1024 ! adpcmdec ! filesink location=dump.pcm",

**Reply by QuantumEntangledAndy** [reply]
That gstreamer caps method that @twistedddx mentioned from #68 would work but only if we striped the block header (with the `0001` and half block size).

**Reply by QuantumEntangledAndy** [reply]
I'm working on a refractor of the BcMedia format to use nom parsers and have proper deserilaization. That should make the code easier to understand and pave the way to writing a serialiser so we can write these talk formats.

**Reply by QuantumEntangledAndy** [reply]
> Isn't the DVI4 block size for this data 516?

Which part of the data are you reading the msg length in the BC header is 516, which is 512 bytes of adpcm plus 4 byte block header with 0001 and u16 of half block size.

**Reply by entropin** [reply]
Fyi: the types involved is: 

201: Length 576 - When record btn is pressed.

202: Length 861 - Continues audio data

11: Length 182 - When button is released

**Reply by twistedddx** [reply]
I read the 01wb
30 31 77 62 08 02 08 02 00 01 02 00 00 00 00 00



01wb BC media header is 8 bytes
30 31 77 62 08 02 08 02
30 31 77 62 = 01wb
size is 08 02 = 520 (repeated twice)

Payload of 01wb is 520 bytes
In those 520 bytes
00 01 02 00 = 4 bytes are second magic and not directly adpcm related
00 00 00 00 = 4 bytes are adpcm state. First 3 bytes here have data as expected if not first 01wb of a packet.
Then 512 byte are adpcm audio.

**Reply by QuantumEntangledAndy** [reply]
Ah yes I see now, I was just reading the `02 00` and forgetting the state is in there too outside of that 512 not inside of 512 bytes.

**Reply by QuantumEntangledAndy** [reply]
There is one adpcm block per `30317762` have we ever seen more than one?

**Reply by twistedddx** [reply]
Never seen it different before.
Every 01wb has 1x second magic and 1x adpcm block.
1x BC protocol packet can contain multiple 01wb/00dc etc.

**Reply by QuantumEntangledAndy** [reply]
@twistedddx Could you help verify that my new BcMedia deserialisation code works on adpcm audio from the camera? It's my `BcMedia_serde` [branch](https://github.com/QuantumEntangledAndy/neolink/tree/BcMedia_serde). It would be appreciated

**Reply by twistedddx** [reply]
Hopefully will have time tomorrow night for that.

**Reply by entropin** [reply]
> I read the 01wb
> 30 31 77 62 08 02 08 02 00 01 02 00 00 00 00 00
> 
> 01wb BC media header is 8 bytes
> 30 31 77 62 08 02 08 02
> 30 31 77 62 = 01wb
> size is 08 02 = 520 (repeated twice)
> 
> Payload of 01wb is 520 bytes
> In those 520 bytes
> 00 01 02 00 = 4 bytes are second magic and not directly adpcm related
> 00 00 00 00 = 4 bytes are adpcm state. First 3 bytes here have data as expected if not first 01wb of a packet.
> Then 512 byte are adpcm audio.

When i parse the date in the cap file  the first two blocks in every BCMedia pack is 516(+4) but the third is always 237(+4). the repeating size bytes is sill saying 520 but there is not roome for that.

Is this "normal" or dose it seams like am doing something wrong?

**Reply by twistedddx** [reply]
I am not infront of the capture right now, but did the data continue in the next packet?

If you stopped the broadcast I imagine it will just abruptly cut the frame short. It likely just appends to the packet as it goes. So header was written with intended length then you as a user told it to stop filling with data, so it did.
Adpcm decoder would play the data it got and then need to tolerate unexpected end of file which is not hard.

**Reply by QuantumEntangledAndy** [reply]
Yes you have to join multiple BC packets up to make complete BcMedia packets. It really is an encapsulated stream it can break up the BcMedia stream anywhere.

I am using a combinator library and a buffered stream reader to do it in neolink.

The combinator library gives hints such as this packet needs 102 bytes to complete which is then requested to the buffered stream reader which pulls BC packets and extracts the binary payloads and appends them to the buffer until it has all 102 bytes.

You might want to see if there's a decent combinator library in c# to handle the heavy lifting.

**Reply by entropin** [reply]
Ok but the first chunk in the next package is again 516 bytes, and not the "remaining" 237 bytes. But i think i get it, just making sure. 

I will post a audio file created from the package capture tomorrow,, i think i have done it all correctly but i still can seam to play it back

**Reply by twistedddx** [reply]
![01wb](https://user-images.githubusercontent.com/12379314/128092685-34d701a0-fc8e-4ce6-a195-af7222147c06.png)

Isn't this the remaining data and more 01wb?
The packets split when it got too big.
Wireshark and the baichuan dissector get weird when a packet splits.
Use Wireshark build in stuff such as the tcp payload to view the data the packets included.

So each of you binary BC Media streams are across 2 packets each due to size.

**Reply by twistedddx** [reply]
![01wb2](https://user-images.githubusercontent.com/12379314/128094234-0ab14b5f-1377-4be7-a6a2-f413b5842669.png)
First packet gets 2 and a half 01wb frames.
Second packet has a half and 1 01wb frames.

Actually Baichuan dissector does appear to show all 4x 01wb.
Eg packet 19 Baichuan dissector has got the payload of packet 18 and 19.

**Reply by QuantumEntangledAndy** [reply]
So playing with my `neolink talk` there was a bug in the `<TalkConfig>` which I have fixed.

I also found it will only accept the `<TalkConfig>` if it matches the `<TalkAbility >` that the camera sends says it can play.

Then I go and try to play the sound out.adpcm it accepts it makes a single "Click" sound and stops....

**Reply by QuantumEntangledAndy** [reply]
Needs more TLC it seems

**Reply by QuantumEntangledAndy** [reply]
So reading the gstreamer [source code](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/blob/master/gst/adpcmenc/adpcmenc.c) for dvi layout in gstreamer there may be an extra byte in the header ....

**Reply by QuantumEntangledAndy** [reply]
I could not find anything different about my packets, except the time they were sent......

So I added a delay to the stream so that new packets would not be sent until they had finished playing on camera and Voila I have sound!!!!

It seems it has no buffer and just plays the sound it gets and if it gets another sound before it finishes playing it just stops.

This was used to encode
```bash
gst-launch-1.0 filesrc location=nero.aac ! decodebin ! audio/x-raw ! audioresample ! "audio/x-raw,rate=16000" ! audioconvert ! "audio/x-raw,channels=1" ! queue  ! adpcmenc blockalign=512 layout=dvi ! filesink location=out.adpcm
```

This was used to play

```bash
neolink talk --config=config.toml --adpcm-file=out.adpcm --block-size=512 --sample-rate=16000 CameraName
```

**Reply by entropin** [reply]
> I could not find anything different about my packets, except the time they were sent......
> 
> So I added a delay to the stream so that new packets would not be sent until they had finished playing on camera and Voila I have sound!!!!
> 

So this to me is pure genius.. I would never have thought to even try this.. The whole process have been very humbling..

So much respect!!

**Reply by QuantumEntangledAndy** [reply]
The newest version of `talk` will play any format of file that gstreamer can understand. Have a go and please let me know if it works

**Reply by QuantumEntangledAndy** [reply]
A have a also added microphone and volume. If I put volume to 0.01 I expect little to no sound but I get clicking in a repeating fashion. I think perhaps something is still up. Maybe the header is not writing correctly or something

**Reply by twistedddx** [reply]
> 
> 
> @twistedddx Could you help verify that my new BcMedia deserialisation code works on adpcm audio from the camera? It's my `BcMedia_serde` [branch](https://github.com/QuantumEntangledAndy/neolink/tree/BcMedia_serde). It would be appreciated

I am not sure the BcMedia_serde is still the branch to test but I tested with a Swann camera with adpcm and it gives errors:

https://pastebin.com/AH1Dq8BP

**Reply by QuantumEntangledAndy** [reply]
Thanks. Could you test on the `talk` branch instead, I found a bug in `BcMedia_serde` where I deserialised the block size incorrectly, thats fixed in `talk` :)

**Reply by twistedddx** [reply]
talk does the same:
https://pastebin.com/LxWY5EWh

**Reply by QuantumEntangledAndy** [reply]
Ok please try again on talk when you can, Iv'e got another fix in

**Reply by twistedddx** [reply]
Different but still error:
https://pastebin.com/Ys8rbgRk

**Reply by QuantumEntangledAndy** [reply]
Interesting I need more debug info on this.  This next build should print the error message of what failed to deserialization

**Reply by twistedddx** [reply]
As a separate thing, earlier you listed:
"gst-launch-1.0 filesrc location=nero.aac ! decodebin ! audio/x-raw ! audioresample ! "audio/x-raw,rate=16000" ! audioconvert ! "audio/x-raw,channels=1" ! queue  ! adpcmenc blockalign=512 layout=dvi ! filesink location=out.adpcm"

I am still fairly confident the native blocksize/blockalign for the packet capture that was given is 516.

A blockalign of 512 will produce 4 bytes of state plus 508 bytes audio. After decode you get 1016 samples.
A blockalign of 516 will produce 4 bytes of state plus 512 bytes audio. After decode you get 1024 samples.
The packet capture shows 4 + 512, at least IMHO.

I have tested with gst-launch-.10 at the different blockalign lengths that it produces the files as I listed above.


The adpcm file gst-launch-1.0 creates with blockalign 516 will at first glance look different to the packet capture.

Packet capture has 8 BCMedia header + 4 magic + 4 state + 512 audio aka 528 bytes which creates a neat 16 byte alignment for hex viewing. But 12 of those bytes are not DVI4 adpcm.
adpcm blockalign of 512 (4+508) also creates a neat 16 byte alignment for hex viewing.
adpcm blockalign of 516 (4+512) does not create a neat alignment for hex viewing, it moves by 4 bytes for each block.

**Reply by QuantumEntangledAndy** [reply]
Hmm I was thinking block size of 512 makes a structure of

4 bytes of header + 512 bytes of samples

This is the source from gst adcpmenc

```c
const int HEADER_SIZE = 4;
gint16 prev_sample[2] = { 0, 0 };
guint32 write_pos = 0;
guint8 channel = 0;
for (channel = 0; channel < enc->channels; channel++) {
    write_pos = channel * HEADER_SIZE;
    outbuf[write_pos + 0] = (samples[channel] & 0xFF);
    outbuf[write_pos + 1] = (samples[channel] >> 8) & 0xFF;
    outbuf[write_pos + 2] = enc->step_index[channel];
    outbuf[write_pos + 3] = 0;
    prev_sample[channel] = samples[channel];
  }

  /* raw-audio looks like this for a stereo stream:
   * [ L, R, L, R, L, R ... ]
   * encoded audio is in eight-sample blocks, two samples to a byte thusly:
   * [ LL, LL, LL, LL, RR, RR, RR, RR ... ] 
   */
  write_pos = HEADER_SIZE * enc->channels;
  read_pos = enc->channels;     /* the first sample is in the header. */
  while (write_pos < enc->blocksize) {
    // Encodes samples
```

And I think I see what you mean that the block size includes the header. I was misreading the `  while (write_pos < enc->blocksize) {` loop thinking that it starts at zero but it actually starts at `  write_pos = HEADER_SIZE * enc->channels;`

**Reply by QuantumEntangledAndy** [reply]
So I went and adjusted the block size and it fixes my clicking issue yay. Now if I can get your adpcm to decode I'll put this in my PR

**Reply by QuantumEntangledAndy** [reply]
Thank you @twistedddx :)

**Reply by twistedddx** [reply]
If anyone wants a wav header for their adpcm files insert the below first line before your 516 bytealign adpcm.
sizes like filesize, fmt size and data size might need to be adjusted smaller if your file was shorter than mine.

```
52494646 A4910000 57415645 666D7420 10000000 11000100 803e0000 00FA7D00 04020400 64617461 ec372200
R I F F  FileSize W A V E  f m t    fmt Size frm chan samples  bytessec blckbsmp d a t a  datasize
```

This should create a file for IMA(1100), mono(0100), samplerate 16000(803e0000), bytesasec 16000*516(00FA7D00), blockalign 516(0402), bitspersample 4 (0400).

foobar, mpc, vlc etc will play the file perfectly.

**Reply by twistedddx** [reply]
Latest build error log:
https://pastebin.com/8FuXn965

**Reply by QuantumEntangledAndy** [reply]
thanks I can see that's it's failing to match the magic. I will need to create a test suite for this I think.

I'll see if I can find some wireshark packets of the stream. I think you sent them in the previous adpcm thread unless youd be kind of enough to send some new ones. I can then use them to verify the decoder myself.

**Reply by twistedddx** [reply]
Packet capture of Neolink trying to connect/reconnect attached.
Btw this particular camera is not in use and was factory defaulted.
Username and password are default for anyone looking ;)
[swann.zip](https://github.com/thirtythreeforty/neolink/files/6931773/swann.zip)

**Reply by QuantumEntangledAndy** [reply]
Thanks once again. I'm off to bed now but I'll put it on my task list for tomorrow.

**Reply by entropin** [reply]
Note, For my camera if the talk session ends before the client has sending the bytes signaling that the session is over (Type 11).
The camera will start responding to TalkConfig package with response_code: 422. Until the camera state is in sync again by sending a type 11. The client seam to always send a  type 11 after response_code: 422

**Reply by QuantumEntangledAndy** [reply]
That's interesting. I haven't observed that but maybe it's because I send logout (2) after a talk session instead. I'll see if I can get a 11 to show up on my own client so I can see it's msg body.

**Reply by QuantumEntangledAndy** [reply]
@twistedddx I have the packets you sent now deserialising correctly could you test for me?

**Reply by entropin** [reply]
> That's interesting. I haven't observed that but maybe it's because I send logout (2) after a talk session instead. I'll see if I can get a 11 to show up on my own client so I can see it's msg body.

Its just an empty XmlExtension.

```
<?xml version="1.0" encoding="UTF-8" ?>
<Extension version="1.1">
<channelId>0</channelId>
</Extension>
```
The camera respondes with the same type and a 200 status if the talk session state ended

**Reply by QuantumEntangledAndy** [reply]
Yes I managed to observe it and its implemented it the current `talk` branch. It now send msg 11 when audio finishes and when receiving 422

**Reply by twistedddx** [reply]
> 
> 
> @twistedddx I have the packets you sent now deserialising correctly could you test for me?

All works except this:
https://github.com/QuantumEntangledAndy/neolink/commit/5952706c08aca9fb5c44f0f356a5ef2d70200c55

These cameras are 8000 sample rate.
At 16000 it plays too fast then stops, then plays too fast etc.
Reverting back to 8000 plays nice.

**Reply by twistedddx** [reply]
Also in https://github.com/QuantumEntangledAndy/neolink/commit/ce1b410f740621436e8b3ea438d2246cadaeebe8
Did you typo 422 and write 442?

**Reply by QuantumEntangledAndy** [reply]
Intesting do you know if the sampling rate is given at any point? The format I need on my camera for talking is 16000, and I thought maybe the 8000 was me thinking its in bytes rather than samples. I can find this sample-rate in the TalkAbility part, I will have a look to see if anything has the number 8000

**Reply by QuantumEntangledAndy** [reply]
55 yes typo it seems

**Reply by QuantumEntangledAndy** [reply]
In your serial xml there is an `<baudRate>9600</baudRate>` do you think it might be that?

**Reply by twistedddx** [reply]
No, 9600 and baud is serial port speed language.
Could you assume 8000 and then overwrite with 16000 when new cameras say it?

**Reply by QuantumEntangledAndy** [reply]
I'm just going to put it to 8000 its only important during decode. If we find a camera with non 8000 we can investigate then

**Reply by QuantumEntangledAndy** [reply]
All done can you have one last try then I'll put this into my PR

**Reply by twistedddx** [reply]
New commit for 8000 works good.

Btw in your typo correction you added 2x comments that include the typo ;)

**Reply by QuantumEntangledAndy** [reply]
Typo fixed. I am going to squash those commits so if you pull this branch again it will probably error. Best to remove and repull the branch

**Reply by entropin** [reply]
Nice, works fine for me! :) I Do have C# code with login/decryption and fully working talk functionality now if anyone is interested, but I think I'll stick with the Rust code for my project. I just have to learn GStreamer.

With C# I feel like I need to fight the bytes at every level, clearly not meant for bit manipulation :) Thanks a thousand for all the help and insight!  I'll be back with a PR for pan/zoom when am feel like I got the basics of Rust down.

**Reply by QuantumEntangledAndy** [reply]
I am glad it works well for you too. It's rather fun to work on this sort of thing. Good luck with the pan and if you need more help decoding you know where to find us :)

**Reply by twistedddx** [reply]
Maybe we(98% your effort @QuantumEntangledAndy ) stole the puzzle win fun from @entropin 

Oops sorry for ruining the game entropion :(

**Comment by entropin** [comment]
**Recompiling the audio files**
So, just to help me fully understand the data I wished to recompile the audio file from the capture, this is my understanding as of now. Please help me make it inline with reality :D

1. I'll take the first 202 pack, find the position of the payload, in this case starting at byte poss 132.
2. Splitting that binary payload stream every time it finds the bytes 01bw (30317762) this will give me 3 chunks of bytes in the case of the first 202 package
3. Removing the first 8 byte from every chunk, to get rid of the BCMedia header (The 01bw is still present on every chunk after the split).
4. Removing 4 more bytes since we don't care about the second magic.

**Q:** Should i also remove the adpcm state? or should that still be in the recompiled audio file?

5, concatenating all the bytes and saving to file (repeat the process and writing more data to the file for every 202 type found in the capture)?

Is this correct ?

Also, do we know of any desktop media player that can play this format (to validate that the file works before sending it back to the camera)?

**Reply by QuantumEntangledAndy** [reply]
Yes I think you have it.

In 4. The second magic dose contain the block size in case you need that, but I think it's in the xml anyways.

You should keep the adpcm state its part of the specification (see [here](https://wiki.multimedia.cx/index.php/Apple_QuickTime_IMA_ADPCM) its for apple ima and but they explain it better in this page then the dvi page)

**Reply by entropin** [reply]
Am sure the answer is in the link you just posted, but am still learning, so o can't find it :D But is there any simple manipulation of the header that i can do with the data, to make it into Ima or the apple dvi format so that i can play the recompiled audio in QuickTime?

**Reply by QuantumEntangledAndy** [reply]
Is there any reason the method you just outlined is not working? I am not sure what else you what by "simple manipulation"

**Reply by entropin** [reply]
The method outlined would produce a DVI4 file that we can feed back to the camera.

But before i do that, i would like to listen to the file, to verify that it works and also to now what output to expect from my camera.

Right now i don't know of a way to listen to the dvi4 formated file on my computer. So if we don't know of a player that can handled that, my thinking is that if i can just switch some header bytes and make it into a format that a normal desktop player can handle, i could do that when I parse the capture file as well.

**Reply by QuantumEntangledAndy** [reply]
Ah yes you mean to listen to it. You could prepend it with the right RIFF wave header...

**Reply by QuantumEntangledAndy** [reply]
I think ffplay should be able to do it without any header tweking too if you have the right command line

**Reply by QuantumEntangledAndy** [reply]
Try this one

```bash
ffplay -f s16le -ar 8000 -acodec adpcm_ima_qt out.adpcm
```

**Reply by QuantumEntangledAndy** [reply]
I also remembered that the BcMedia has a padding so that its divisible by 8. So if your payload (in the BcMedia heading) is 12 then the adpcm data is 4..11 then there is a padding from 12..15 (0 based) before the next BcMedia starts.

```
00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15
HH HH HH HH AD AD AD AD AD AD AD AD PP PP PP PP
```

HH Magic and block size header

AD Adpcm data

PP Padding (just zero bytes)

**Reply by entropin** [reply]
> Try this one
> 
> ```bash
> ffplay -f s16le -ar 8000 -acodec adpcm_ima_qt out.adpcm
> ```

This expect the apple QuickTime version of adpcm, what magic is s16le? :D  (will test when am done with the packade parser code)

**Reply by QuantumEntangledAndy** [reply]
s16le = signed 16 bit little endian

**Reply by QuantumEntangledAndy** [reply]
So I am not sure if your aware but in #185 I have been adding subcommand to neolink like

```bash
neolink reboot --config=config.toml CameraName
```

I also in my `BcMedia_serde` branch I have been working on making BcMedia packets easily  de/serializable. Using these two I think I can write some code to add this talk functionality into neolink.

**Reply by entropin** [reply]
That would be awesome, my playground code is close to working i think, but its still only in c#  Still a long way to learn before i can create a Rust PR of any quality

**Reply by QuantumEntangledAndy** [reply]
So I have my first iteration of the `neolink talk` up. Needs testing!

I am going to get some food now but if you want to play with it check out my `talk` branch on my [fork](https://github.com/QuantumEntangledAndy/neolink). 

Once complied run

```bash
neolink help talk
```

To see the options then try something like

```bash
neolink talk --config=config.toml --adpcm-file=data.adpcm --sample-rate=16000 --block-size=512 NameOfCamera
```

I make no guarantees that it is working in any shape this is just my first quick workup

**Reply by twistedddx** [reply]
For playback you could use gstreamer to convert the file back to s16le pcm inside wav container.

gst-launch-1.0 filesrc location=dump.adpcm  ! caps=audio/x-adpcm,layout=dvi,block_align=516,channels=1,rate=16000  ! adpcmdec ! wavenc ! filesink location=dump.wav

PS I didn't check the above and it may need some love.

**Reply by QuantumEntangledAndy** [reply]
Yeah that's a good idea. When I've gotten it working with raw bytes I'll look into hooking it up to gstreamer and get it to convert.

**Reply by twistedddx** [reply]
you likely would want to use adpcmenc to create this adpcm data from more common formats.
Eg
Feed Neolink a wav, aac or MP3.
Neolink uses gstreamer to convert to DVI adpcm

**Reply by QuantumEntangledAndy** [reply]
Yes see if I can get it to stream from the microphone or stdin too.

Also that will let us match the format to exactly what TalkAbility reports is the best format.

**Reply by entropin** [reply]
Oki, so is still having problems playing back the files created from the dump on my computer, the files generated seams fine and all bytes lines up. So I have a few questions.

This command generates the adpcm file. Is the consensus that this now should be blockalign=516? and not 512?

 `gst-launch-1.0 filesrc location=data.aac ! decodebin ! audio/x-raw ! audioresample ! "audio/x-raw,rate=16000" ! audioconvert ! "audio/x-raw,channels=1" ! queue  ! adpcmenc blockalign=512 layout=dvi ! filesink location=out.adpcm`

Anyhow, how do I simply reverse this command? Turning the file back to aac (gst documentation is wast.., bin stuck her for a while).

@twistedddx You mentioned wave headers, can I just stick that in the begining of to the out.adpcm file and play it in vlc? Or dose the file out.adpcm have a header that i need to replace?

**Reply by entropin** [reply]
the wave header you mentioned worked like a charm on my recompiled audio, than you!
52494646 A4910000 57415645 666D7420 10000000 11000100 803e0000 00FA7D00 04020400 64617461 ec372200

**Reply by QuantumEntangledAndy** [reply]
Yes consensus is adpcm is 516 block-align that seems to work and give good audio on the camera while 512 did not.

This is my encoding

```bash
gst-launch-1.0 filesrc location=target/aac/nero.aac ! decodebin ! audioconvert ! audioresample ! "audio/x-raw,rate=16000,channels=1" ! volume volume=2.0 ! adpcmenc blockalign=516 layout=dvi ! filesink location=target/aac/out.adpcm
```

Playing it from the out.adpcm is difficult. It lacks a header that describes the data so most progams like gst and ffmpeg cannot deduce it's format.

You can try this one

```
gst-launch-1.0 filesrc location=target/aac/out.adpcm ! typefind force-caps=audio/x-adpcm,layout=dvi,block_align=516,channels=1,rate=16000 ! adpcmdec ! audioconvert ! autoaudiosink
```

Which should use the `force-caps` part of typefind to set the format but I have never gotten it to work. :(
